docs/components/authentication.md

```diff
@@ -0,0 +1,185 @@
+with authentication detail (for example by adding an `Authorization` header).
+This is useful when you have to send multiple requests to the same endpoint. Using an authentication implementation,
+these details can be separated from the actual requests.
+
+
+## Installation
+
+```
+$ composer require php-http/message
+```
+
+
+## Authentication methods
+
+General usage looks like the following:
+
+``` php
+$authentication = new AuthenticationMethod();
+
+/** @var Psr\Http\Message\RequestInterface */
+$authentication->authenticate($request);
+```
+
+
+### Basic Auth
+
+This authentication method accepts two parameters: username and password. Getters/Setters are provided by
+`Http\Message\Authentication\UserPasswordPair` trait.
+
+``` php
+use Http\Message\Authentication\BasicAuth;
+
+$authentication = new BasicAuth('username', 'password');
+
+// These details can be set later as well
+// There are also getters with the appropriate method names
+$authentication->setUsername('username');
+$authentication->setPassword('password');
+```
+
+
+### Bearer
+
+This authentication method accepts one parameter: a token.
+
+``` php
+use Http\Message\Authentication\Bearer;
+
+$authentication = new Bearer('token');
+
+// Token can be set later as well
+$authentication->setToken('token');
+```
+
+
+### Chain
+
+This authentication method accepts one parameter: an array of other authentication methods.
+
+The idea behind this authentication method is that in some cases you might need to
+authenticate the request with multiple methods.
+
+For example it's a common practice to protect development APIs with Basic Auth and the regular token auth as well
+to protect the API from unnecessary processing.
+
+
+``` php
+use Http\Message\Authentication\Chain;
+
+$authenticationChain = [
+    new AuthenticationMethod1(),
+    new AuthenticationMethod2(),
+];
+
+$authentication = new Chain($authenticationChain);
+
+// Authentication chain can be modified later
+$authenticationChain = $authentication->getAuthenticationChain();
+
+array_pop($authenticationChain);
+
+$authentication->setAuthenticationChain($authenticationChain);
+
+$authentication->clearAuthenticationChain();
+
+$authentication->addAuthentication(new AuthenticationMethod3());
+```
+
+
+### Matching
+
+With this authentication method you can conditionally add authentication details to your request by passing a callable
+to it. When a request is passed, the callable is called and used as a boolean value in order to decide whether
+the request should be authenticated or not.
+It also accepts an authentication method instance which does the actual authentication when the condition is
+fulfilled.
+
+For example a common use case is to authenticate requests sent to certain paths.
+
+
+``` php
+use Http\Message\Authentication\Mathing;
+
+$authentication = new Mathing(new AuthenticationMethod1(), function () { return true; });
```

Comment:
Mathing or Matching?

i would make the example a real one, matching for only some domain. then we would see how the matching callback is passed the request (i guess?) and does a real check on it.